homeassistant/components/homekit_controller/connection.py

```diff
@@ -79,27 +78,77 @@ def __init__(self, hass, host, port, model, hkid, config_num, config):
 
         hass.data[KNOWN_DEVICES][hkid] = self
 
-        if self.pairing is not None:
-            self.accessory_setup()
-        else:
+    def setup(self):
+        """Prepare to use a paired HomeKit device in homeassistant."""
+        if self.pairing is None:
             self.configure()
+            return
+
+        self.pairing.pairing_data['AccessoryIP'] = self.host
+        self.pairing.pairing_data['AccessoryPort'] = self.port
+
+        cache = self.hass.data[ENTITY_MAP].get_map(self.unique_id)
+        if not cache or cache['config_num'] < self.config_num:
+            return self.refresh_entity_map(self.config_num)
+
+        self.accessories = cache['accessories']
 
-    def accessory_setup(self):
+        # Ensure the Pairing object has access to the latest version of the
+        # entity map.
+        self.pairing.pairing_data['accessories'] = self.accessories
+
+        self.add_entities()
+
+        return True
+
+    async def async_refresh_entity_map(self, config_num):
+        """
+        Handle setup of a HomeKit accessory.
+
+        The sync version will be removed when homekit_controller migrates to
+        config flow.
+        """
+        return await self.hass.async_add_executor_job(
+            self.refresh_entity_map,
+            config_num,
+        )
+
+    def refresh_entity_map(self, config_num):
         """Handle setup of a HomeKit accessory."""
         # pylint: disable=import-error
-        from homekit.model.services import ServicesTypes
         from homekit.exceptions import AccessoryDisconnectedError
 
-        self.pairing.pairing_data['AccessoryIP'] = self.host
-        self.pairing.pairing_data['AccessoryPort'] = self.port
-
         try:
-            data = self.pairing.list_accessories_and_characteristics()
+            accessories = self.pairing.list_accessories_and_characteristics()
         except AccessoryDisconnectedError:
-            call_later(
-                self.hass, RETRY_INTERVAL, lambda _: self.accessory_setup())
+            # If we fail to refresh this data then we will naturally retry
+            # later when Bonjour spots c# is still not up to date.
             return
-        for accessory in data:
+
+        self.hass.data[ENTITY_MAP].async_create_or_update_map(
```

Comment:
Don't call async functions from sync context and vice versa.

Reply:
It's not actually a coroutine, but it does call `store.async_delay_save` and is, I believe, 'async safe'. So I could mark it with @callback or drop the async_ prefix?

Reply:
Async callbacks should also only be executed in the event loop.